pkgs/development/compilers/polyml/default.nix

```diff
@@ -25,6 +25,11 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = lib.optional stdenv.isDarwin autoreconfHook;
 
+  preConfigure = ''
+    # Needed to fix g++'s path in polyc
+    export CXX="$(command -v ${stdenv.cc.targetPrefix}c++)"
```

Comment:
How is that supposed to work when cross-compiling?

I think the package should be using `gcc` or `clang` in build inputs and fix whatever logic embeds build-time C++ compiler as a default for runtime.

Reply:
> How is that supposed to work when cross-compiling? The host platform may not be able to execute binaries for the build platform.

The `${stdenv.cc.targetPrefix}c++` command refers to the stdenv compiler for the host platform, not the build platform.

And as far as I can see, the `polyc` script is not used during the build. However, I cannot be sure of that, because cross-compilation of `polyml` is broken even without this PR.

> in any case, `$NIX_CC/bin/$CXX` instead of `command -v …`

I'd rather use `${stdenv.cc.targetPrefix}`, which [is documented in the nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#ssec-cross-cookbook) (in the cross-packaging cookbook, even), than `$NIX_CC`, which is undocumented.

> also use type -P instead of `command -v` for PATH search

OK, I'll change that.

Reply:
> The ${stdenv.cc.targetPrefix}c++ command refers to the stdenv compiler for the host platform, not the build platform.

Sorry, I think I got confused, and it's the other way around. I'll see how I can fix this (possibly with `substituteInPlace` on the `polyc` script, rather than overriding `$CXX` so as to not affect the build).

Reply:
>However, I cannot be sure of that, because cross-compilation of polyml is broken even without this PR.

Fair enough, but it’s relatively easy to fix (i.e. convince generated `./polyimport`, `./poly` wrapper scripts to use an emulator).

```console
$ uname -m
aarch64
$ file $out/bin/poly
[REDACTED]/nixpkgs/outputs/out/bin/poly: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/v74n672akwg6pc3cjhd2lr3bakdfvb2m-glibc-x86_64-unknown-linux-gnu-2.39-52/lib/ld-linux-x86-64.so.2, for GNU/Linux 3.10.0, not stripped
```

>possibly with substituteInPlace on the polyc script

Perhaps replace `@CXX@` in postPatch with either unambiguous variable set from `configure.ac`, or simply replace with path to the C++ compiler.

>The ${stdenv.cc.targetPrefix}c++ command refers to the stdenv compiler for the host platform, not the build platform.

`targetPrefix` doesn’t refer to anything on its own, it’s an alias for platform’s `config` attribute.

https://github.com/NixOS/nixpkgs/blob/b098e0c792c4aa5af5b8b698b840383f045f8ab7/pkgs/build-support/cc-wrapper/default.nix#L96

In this case, `${stdenv.cc.targetPrefix}c++` is the program name of the C++ compiler for host platform that **runs on build platform**. However, PolyML needs a compiler that **runs on host platform** (and I assume **builds for target platform**, i.e. buildInputs, host → target, see https://nixos.org/manual/nixpkgs/stable/#ssec-stdenv-dependencies-propagated).